src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportFileManager.cs

```diff
@@ -0,0 +1,134 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Health.Fhir.Core.Features.Operations.Export.ExportDestinationClient;
+using Microsoft.Health.Fhir.Core.Features.Operations.Export.Models;
+
+namespace Microsoft.Health.Fhir.Core.Features.Operations.Export
+{
+    /// <summary>
+    /// Helper class that takes care of managing files for export. Currently it takes
+    /// care of creating new files (at the beginning of export as well as once a file
+    /// has reached a certain size) and keeping track of current file for each resource type.
+    /// Expected use case: each ExportJobTask will instantiate a copy; single-threaded access.
+    /// </summary>
+    internal class ExportFileManager
+    {
+        private readonly ExportJobRecord _exportJobRecord;
+        private readonly IExportDestinationClient _exportDestinationClient;
+        private readonly IDictionary<string, ExportFileInfo> _resourceTypeToFileInfoMapping;
+        private readonly uint _maxFileSizeInBytes;
+        private bool _isInitialized = false;
+
+        public ExportFileManager(ExportJobRecord exportJobRecord, IExportDestinationClient exportDestinationClient)
+        {
+            EnsureArg.IsNotNull(exportJobRecord, nameof(exportJobRecord));
+            EnsureArg.IsNotNull(exportDestinationClient, nameof(exportDestinationClient));
+            EnsureArg.IsGt(exportJobRecord.MaxFileSizeInMB, 0, nameof(exportJobRecord.MaxFileSizeInMB));
+
+            _exportJobRecord = exportJobRecord;
+            _exportDestinationClient = exportDestinationClient;
+            _maxFileSizeInBytes = _exportJobRecord.MaxFileSizeInMB * 1024 * 1024;
+            _resourceTypeToFileInfoMapping = new Dictionary<string, ExportFileInfo>();
+        }
+
+        private async Task Initialize(CancellationToken cancellationToken)
+        {
+            // Each resource type can have multiple files. We need to keep track of the latest file.
+            foreach (KeyValuePair<string, List<ExportFileInfo>> output in _exportJobRecord.Output)
+            {
+                ExportFileInfo latestFile = output.Value[output.Value.Count - 1];
```

Comment:
I'm concerned about the order the files would be in after a restart, whether the one with the highest sequence will always be the last one in the array.

Reply:
My understanding was that since we are storing the ExportFileInfo as a list, the order should be preserved when de-serializing the json into an object. But I looked at the documentation and it does not specifically mention that order will be preserved (though practically that seems to be the case). To be on the safer side, I have updated the code to loop through the export files and find the most recent one.